### PR TITLE
Generalize palautelomake

### PR DESCRIPTION
**Pull request by mremes**
This PR adds a framework to create forms which will be usable from the site and submitting them will send results to given recipients. Instead of creating custom handlers for each form, one just needs to implement the following objects:

* callable that returns a `ContactForm` instance,
* Jinja template deriving from the base `palautteet.html` template and
* email details

See examples in `palautteet/hallitus/views.py` (replacing the existing `hallituspalaute` form) and `palautteet/hairinta/views.py` (a new form). Later these resources can be autogenerated with use of some kind of a form builder and form to fill in the details.

Please give some feedback if you have any.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,7 @@ build:
 	docker tag ${IMG} ${LATEST}
 
 test:
-	docker run ${NAME} /bin/sh -c "python manage.py test"
+	docker run ${NAME} /bin/sh -c "python manage.py test ilmo_app/ palautteet/tests"
```
Comment (vkopio) [inline Makefile]: The former command should have already ran all the tests in every app if they are named correctly.

File: palautteet/email.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from abc import ABC, abstractmethod
+from typing import Mapping, List
+
+from django.core.mail import EmailMessage
+
+DEFAULT_SENDER = 'noreply@satakuntalainenosakunta.fi'
```
Comment (vkopio) [inline palautteet/email.py]: satakuntatalo.fi is used in every other email.

File: palautteet/hairinta/static/js/conditional_fields.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    if (document.getElementsByName(checkboxName)[0].checked) {
+        display = "inline";
+    } else {
+        fields.forEach((x) => document.getElementsByName(x)[0].value = "");
+        display = "none";
+    }
+    divs.forEach((x) => document.getElementsByName(x).forEach((y) => y.style.display=display));
+    fields.forEach((x) => document.getElementsByName(x).forEach((y) => y.style.display=display));
+
+}
```
Comment (vkopio) [inline palautteet/hairinta/static/js/conditional_fields.js]: No new line at the end.

File: palautteet/hairinta/templates/hairintailmoitus.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    Häirintäilmoitus
+{% endblock %}
+
+{% block header %}
+    Häirintäilmoitus
+{% endblock %}
+
+{% block thank_you_text %}
+    Kiitos ilmoituksesta!
+{% endblock %}
```
Comment (vkopio) [inline palautteet/hairinta/templates/hairintailmoitus.html]: No new line at the end.

File: palautteet/hallitus/templates/hallituspalaute.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        Olet kirjautunut sisään. Avaa <a href="">tämä sivu</a> incognito/private-tilassa, jos haluat lähettää nimettömän
+        viestin.
+    {% else %}
+        <a href="/login">Kirjaudu sisään</a>, jos haluat jättää kirjallisen aloitteen.
+    {% endif %}
+{% endblock %}
+
+{% block thank_you_text %}
+    Kiitos viestistäsi! Vastaamme siihen mahdollisimman pian, mikäli niin halusit.
+{% endblock %}
```
Comment (vkopio) [inline palautteet/hallitus/templates/hallituspalaute.html]: No new line at the end.

File: palautteet/templates/palautteet.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                    {% endfor %}
+
+                    <input type="submit" class="btn btn-primary" value="Lähetä"/>
+                </form>
+            {% endblock %}
+        {% elif ok %}
+            <p>{% block thank_you_text %}{% endblock %}</p>
+        {% endif %}
+    </div>
+{% endblock %}
```
Comment (vkopio) [inline palautteet/templates/palautteet.html]: No new line at the end.

File: palautteet/views.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    if not form.is_valid():
+        return HttpResponseBadRequest('Invalid form data')
+    _send_email(form, service)
+    return render(request, template_name, dict(ok=True))
+
+
+def get_form_view(form_fn: Union[Callable[[User], ContactsForm],
+                                 Callable[[User,  dict], ContactsForm]],
+                  template_name: str,
+                  service: AbstractEmailService):
```
Comment (vkopio) [inline palautteet/views.py]: The attributes could be listed a bit more cleanly as described above

File: palautteet/email.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    @abstractmethod
+    def send(self, fields: Mapping[str, object], reply_to: str = None):
+        pass
+
+
+class EmailService(AbstractEmailService):
+    def __init__(self,
+                 subject: str,
+                 from_email: str,
+                 to_email: List[str]):
```
Comment (vkopio) [inline palautteet/email.py]: The list of parameters seems to be short enough to be on the same line.

File: palautteet/hairinta/views.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+from ..email import EmailService
+from ..views import get_form_view
+from .forms import HairintaForm
+
+
+TEMPLATE = 'hairintailmoitus.html'
+EMAIL_SERVICE = EmailService('Häirintäilmoitus',
+                             'hairinta@satakuntatalo.fi',
+                             ['kuraattori@satakuntatalo.fi'])
```
Comment (vkopio) [inline palautteet/hairinta/views.py]: If the list of parameters gets too large, I have started the list from a new line. If the name of the function changes, the other lines don't have to be realigned.
```
EMAIL_SERVICE = EmailService(
    'Häirintäilmoitus',
    'hairinta@satakuntatalo.fi',
    ['kuraattori@satakuntatalo.fi']
)
```

File: palautteet/hairinta/views.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+
+hairintailmoitus = get_form_view(lambda _=None, data=None: HairintaForm(data=data),
+                                 TEMPLATE,
+                                 EMAIL_SERVICE)
```
Comment (vkopio) [inline palautteet/hairinta/views.py]: See above comment. Also, every parameter should be at its own line for clarity.

File: palautteet/hallitus/forms.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+class PalauteFormLogged(PalauteForm):
+    aloite = ChoiceField(
+        choices=[(True, 'Kyllä'), (False, 'Ei')],
+        widget=RadioSelect,
+        label=("Olen osakunnan varsinainen jäsen ja tämä on kirjallinen "
+               "aloite, jonka haluan hallituksen käsittelevän."),
+        help_text=("Osakunnan säännöt §45: Osakunnan varsinaisella jäsenellä "
+                   "on oikeus tekemällään kirjallisella aloitteella saattaa "
+                   "hallituksen toimivaltaan kuuluva asia sen käsiteltäväksi.")
```
Comment (vkopio) [inline palautteet/hallitus/forms.py]: See above comments.

File: palautteet/hallitus/views.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+TEMPLATE = 'hallituspalaute.html'
+EMAIL_SERVICE = EmailService('Postia SatO:n hallitukselle',
+                             'hallituspalaute@satakuntatalo.fi',
+                             ['hallitus@satakuntatalo.fi'])
+
+
+hallituspalaute = get_form_view(get_palaute_form,
+                                TEMPLATE,
+                                EMAIL_SERVICE)
```
Comment (vkopio) [inline palautteet/hallitus/views.py]: See above comments.

File: palautteet/hallitus/views.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+from .utils import get_palaute_form
+from ..email import EmailService
+from ..views import get_form_view
+
+TEMPLATE = 'hallituspalaute.html'
+EMAIL_SERVICE = EmailService('Postia SatO:n hallitukselle',
+                             'hallituspalaute@satakuntatalo.fi',
+                             ['hallitus@satakuntatalo.fi'])
```
Comment (vkopio) [inline palautteet/hallitus/views.py]: See above comments.

File: palautteet/hairinta/views.py (hunk 9 above)
Comment (mremes) [reply palautteet/hairinta/views.py]: Good point. We should get a linter (and an autoformatter such as `black` as well) :P

**Pull request merged**